instrumentation/rocketmq-client-4.8/testing/src/main/groovy/io/opentelemetry/instrumentation/rocketmq/AbstractRocketMqClientTest.groovy

```diff
@@ -85,10 +85,8 @@ abstract class AbstractRocketMqClientTest extends InstrumentationSpecification {
             "$SemanticAttributes.MESSAGING_SYSTEM" "rocketmq"
             "$SemanticAttributes.MESSAGING_DESTINATION" sharedTopic
             "$SemanticAttributes.MESSAGING_DESTINATION_KIND" "topic"
-            "$SemanticAttributes.MESSAGING_MESSAGE_ID" String
             "messaging.rocketmq.tags" "TagA"
             "messaging.rocketmq.broker_address" String
-            "messaging.rocketmq.send_result" "SEND_OK"
```

Comment:
does this mean that you were able to repro the (or at least "a") problem?

Reply:
That part I seriously don't understand - those assertions have clearly passed before, but when I added these null checks they suddenly started failing with missing attributes.
Maybe there are some internal try-catch retries at work there? No idea honestly.

Reply:
Had a quick peek, seems to be caused by the NPE fix in `messageId(...)`. The unexpected thing here is that `TracingSendMessageHookImpl.sendMessageAfter` is called twice. Once before send result is set and second time after. Before this change the code flow was so that it got NPE at the first call when extracting message id and succeeded on the second call. Maybe we should change the condition at the start of `TracingSendMessageHookImpl.sendMessageAfter` to `if (context == null || (context.getSendResult() == null && context.getException() == null)) {` also should pass the value of `context.getException()` to `instrumenter.end(...)`.

Reply:
> The unexpected thing here is that `TracingSendMessageHookImpl.sendMessageAfter` is called twice.

Hm, is it always called twice? It looks like it's just called in a retry loop: https://github.com/apache/rocketmq/blob/master/client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java#L887

> also should pass the value of `context.getException()` to `instrumenter.end(...)`.

Yeah that's a great idea 👍

Reply:
I have no idea whether it is always called twice. Maybe only for async requests? Idk. For `test rocketmq produce callback` first call is from
```
	at io.opentelemetry.javaagent.shaded.instrumentation.rocketmq.TracingSendMessageHookImpl.sendMessageAfter(TracingSendMessageHookImpl.java:47)
	at org.apache.rocketmq.client.impl.producer.DefaultMQProducerImpl.executeSendMessageHookAfter(DefaultMQProducerImpl.java:962)
	at org.apache.rocketmq.client.impl.producer.DefaultMQProducerImpl.sendKernelImpl(DefaultMQProducerImpl.java:871)
	at org.apache.rocketmq.client.impl.producer.DefaultMQProducerImpl.sendDefaultImpl(DefaultMQProducerImpl.java:584)
	at org.apache.rocketmq.client.impl.producer.DefaultMQProducerImpl.access$300(DefaultMQProducerImpl.java:97)
	at org.apache.rocketmq.client.impl.producer.DefaultMQProducerImpl$4.run(DefaultMQProducerImpl.java:511)
```

Second call is from
```
	at io.opentelemetry.javaagent.shaded.instrumentation.rocketmq.TracingSendMessageHookImpl.sendMessageAfter(TracingSendMessageHookImpl.java:47)
	at org.apache.rocketmq.client.impl.producer.DefaultMQProducerImpl.executeSendMessageHookAfter(DefaultMQProducerImpl.java:962)
	at org.apache.rocketmq.client.impl.MQClientAPIImpl$1.operationComplete(MQClientAPIImpl.java:549)
	at org.apache.rocketmq.remoting.netty.ResponseFuture.executeInvokeCallback(ResponseFuture.java:54)
	at org.apache.rocketmq.remoting.netty.NettyRemotingAbstract$2.run(NettyRemotingAbstract.java:319)
```

Reply:
Yeah it looks like it's only happening in async calls. I'll try using the modified condition you posted.